src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/TestUriHelper.cs

```diff
@@ -11,39 +11,68 @@ public static class TestUriHelper
     {
         public static Uri BuildTestUri()
         {
-            return new UriBuilder("http", "localhost", GetNextPort()).Uri;
+            return BuildTestUri(null);
         }
 
         public static Uri BuildTestUri(string hint)
+        {
+            return BuildTestUri(hint, statusMessagesEnabled: false);
+        }
+
+        public static Uri BuildTestUri(string hint, bool statusMessagesEnabled)
         {
             if (string.IsNullOrEmpty(hint))
             {
-                return BuildTestUri();
+                if (statusMessagesEnabled)
+                {
+                    // Most functional tests use this codepath and should directly bind to dynamic port "0" and scrape
+                    // the assigned port from the status message, which should be 100% reliable since the port is bound
+                    // once and never released.  Binding to dynamic port "0" on "localhost" (both IPv4 and IPv6) is not
+                    // supported, so the port is only bound on "127.0.0.1" (IPv4).  If a test explicitly requires IPv6,
+                    // it should provide a hint URL with "localhost" (IPv4 and IPv6) or "[::1]" (IPv6-only).
+                    return new UriBuilder("http", "127.0.0.1", 0).Uri;
```

Comment:
Why default to v4 rather than v6?

Reply:
If the test doesn't care, I assume IPv4 is safer since some environments don't support IPv6 (parts of Azure, maybe Travis/AppVeyor).